src/Validator/Authentication.php

```diff
@@ -189,61 +200,71 @@ public function setService(AuthenticationService $service)
     {
         if ($value !== null) {
             $this->setCredential($value);
         }
 
+        if ($this->identity === null) {
+            throw new Exception\RuntimeException('Identity must be set prior to validation');
+        }
         if (($context !== null) && array_key_exists($this->identity, $context)) {
             $identity = $context[$this->identity];
         } else {
             $identity = $this->identity;
         }
-        if (!$this->identity) {
-            throw new Exception\RuntimeException('Identity must be set prior to validation');
-        }
 
+        if ($this->credential === null) {
+            throw new Exception\RuntimeException('Credential must be set prior to validation');
+        }
         if (($context !== null) && array_key_exists($this->credential, $context)) {
             $credential = $context[$this->credential];
         } else {
             $credential = $this->credential;
         }
 
-        if (!$this->adapter) {
-            throw new Exception\RuntimeException('Adapter must be set prior to validation');
-        }
-        $this->adapter->setIdentity($identity);
-        $this->adapter->setCredential($credential);
-
         if (!$this->service) {
             throw new Exception\RuntimeException('AuthenticationService must be set prior to validation');
         }
+
+        if (!$this->adapter) {
+            $adapter = $this->service->getAdapter();
+            if (!$adapter) {
+                throw new Exception\RuntimeException('Adapter must be set prior to validation');
+            }
+            if (!$adapter instanceof ValidatableAdapterInterface) {
+                throw new Exception\RuntimeException(sprintf(
+                    'Adapter must be an instance of ValidatableAdapterInterface, %s given',
+                    (is_object($adapter) ? get_class($adapter) : gettype($adapter))
+                ));
+            }
+        } else {
+            $adapter = $this->adapter;
+        }
+
+        $adapter->setIdentity($identity);
+        $adapter->setCredential($credential);
+
         $result = $this->service->authenticate($this->adapter);
 
-        if ($result->getCode() != Result::SUCCESS) {
-            switch ($result->getCode()) {
-                case Result::FAILURE_IDENTITY_NOT_FOUND:
-                    $this->error(self::IDENTITY_NOT_FOUND);
-                    break;
-                case Result::FAILURE_CREDENTIAL_INVALID:
-                    $this->error(self::CREDENTIAL_INVALID);
-                    break;
-                case Result::FAILURE_IDENTITY_AMBIGUOUS:
-                    $this->error(self::IDENTITY_AMBIGUOUS);
-                    break;
-                case Result::FAILURE_UNCATEGORIZED:
-                    $this->error(self::UNCATEGORIZED);
-                    break;
-                default:
-                    $this->error(self::GENERAL);
+        if (!$result->isValid()) {
```

Comment:
Flip this conditional around:

```php
if $result->isValid()) {
    return true;
}
// now do error work
```

Return early whenever you can, as it helps make the main code flow more readable.

I'll do this in a separate commit I push to your branch.